mlflow/store/artifact/unity_catalog_models_artifact_repo.py

```diff
@@ -0,0 +1,114 @@
+import json
+
+from mlflow.exceptions import MlflowException
+from mlflow.protos.databricks_uc_registry_messages_pb2 import (
+    GenerateTemporaryModelVersionCredentialsResponse,
+    MODEL_VERSION_READ_WRITE,
+)
+from mlflow.protos.databricks_pb2 import INVALID_PARAMETER_VALUE
+from mlflow.store.artifact.artifact_repo import ArtifactRepository
+from mlflow.utils.databricks_utils import get_databricks_host_creds
+from mlflow.utils.rest_utils import call_endpoint
+from mlflow.utils.uri import (
+    get_databricks_profile_uri_from_artifact_uri,
+    get_db_info_from_uri,
+    is_databricks_unity_catalog_uri,
+    _DATABRICKS_UNITY_CATALOG_SCHEME,
+)
+from mlflow.store.artifact.utils.models import (
+    get_model_name_and_version,
+)
+
+from mlflow.store._unity_catalog.registry.utils import get_artifact_repo_from_storage_info
+
+REGISTRY_GET_SCOPED_TOKEN_ENDPOINT = (
+    "/mlflow/unity-catalog/model-versions/generate-temporary-credentials"
+)
+
+
+class UnityCatalogModelsArtifactRepository(ArtifactRepository):
+    """
+    Performs storage operations on artifacts controlled by a Unity Catalog model registry
+
+    Temporary scoped tokens for the appropriate cloud storage locations are fetched from the
+    remote backend and used to download model artifacts.
+
+    The artifact_uri is expected to be of the form
+    - `models:/<model_name>/<model_version>`
+    - `models:/<model_name>/<stage>`  (refers to the latest model version in the given stage)
+    - `models:/<model_name>/latest`  (refers to the latest of all model versions)
+    - `models://<profile>/<model_name>/<model_version or stage or 'latest'>`
+
+    Note : This artifact repository is meant is to be instantiated by the ModelsArtifactRepository
+    when the client is pointing to a Unity Catalog model registry.
+    """
+
+    def __init__(self, artifact_uri, registry_uri):
+        if not is_databricks_unity_catalog_uri(registry_uri):
+            raise MlflowException(
+                message="Attempted to instantiate an artifact repo to access models in the "
+                f"Unity Catalog with non-Unity Catalog registry URI '{registry_uri}'. "
+                f"Please specify a Unity Catalog registry URI of the "
+                f"form '{_DATABRICKS_UNITY_CATALOG_SCHEME}[://profile]', e.g. by calling "
+                f"mlflow.set_registry_uri('{_DATABRICKS_UNITY_CATALOG_SCHEME}') if using the "
+                f"MLflow Python client",
+                error_code=INVALID_PARAMETER_VALUE,
+            )
+        super().__init__(artifact_uri)
+        from mlflow.tracking.client import MlflowClient
+
+        registry_uri_from_artifact_uri = get_databricks_profile_uri_from_artifact_uri(
```

Comment:
If the models:/ artifact URI itself specifies a Databricks profile via syntax like `models:/dev_profile@databricks-uc/mymodelname/3`, extract the underlying Databricks profile URI (e.g. `databricks-uc://dev_profile`) and use it as the registry URI